### PR TITLE
Forbid remote URLs for avatars/banners (fixes #1618)

### LINKED ISSUE #1618: No validation for avatars source
### Issue Summary

Users can put any url in their avatar and banner. By using it to pass their own website they can get information about IP addresses of other users as soon as they visit lemmy instance. There is also no check of the link, so it could introduce some other problems eg when user will use some big file.

### Steps to Reproduce

1. Send "SaveUserSettings" ws request with chosen url in "avatar" or "banner" field

### Proposal

Check if the host of image is the same as instance host and maybe if there is "pictrs/image" in url

### PR DESCRIPTION
**Pull request by Nutomic**

### CODE REVIEW

**Comment by dessalines** [comment]
This would break federated avatars and banners, which we do allow to be remote: Ex: https://lemmy.ml/c/genzedong@lemmygrad.ml

Why is it necessary to have potentially hundreds of copies of these images on every instance? I don't think the security implications of seeing an IP viewing a picture are nearly as negative as wasting hard drive space by pointlessly duplicating images tons of times.

We need a decentralized picture store that shares hosting costs, rather than continuing this same problem of having every platform duplicate and host the same images sometimes 9 times. The duplication is fine for tiny info like text, but we really shouldn't allow it for resource-constrained self-hosted servers.

If we did add something like this, it'd have to be after we choose to cache these locally.

**Comment by Nutomic** [comment]
No it doesnt break that. The icon/banner on that page are loaded directly from lemmygrad.ml, as you can check for yourself with browser dev tools. And the federation checks verify exactly that, that images are from the same domain as the actor (so for a lemmygrad.ml community, icon/banner have to be hosted on lemmygrad.ml as well). This is very unlikely to break anything, because lemmy-ui only provides an upload button for images, no raw url input. So the only way to have images which fail these checks would be with raw api calls (or with a different client).

I have no idea how you get the idea that there are hundreds of copies of these images. Again, they are loaded directly from the original instance, without being copied onto any other instance. Maybe you are thinking of Mastodon which caches remote images for privacy reasons.

File: crates/api_common/src/lib.rs [hunk 1]
```diff
@@ -629,3 +630,15 @@ pub async fn remove_user_data_in_community(
 
   Ok(())
 }
+
+pub fn check_image_has_local_domain(url: &Option<String>) -> Result<(), LemmyError> {
+  if let Some(url) = url {
+    let settings = Settings::get();
+    let url = Url::parse(url)?;
+    let domain = url.domain().expect("url has domain");
+    if domain != settings.hostname {
```
[the remark this answers is not in the document]
Comment (Nutomic) [reply crates/api_common/src/lib.rs:639]: Yes, and that check is only used when a user/site/community is created/edited through the api. So the image domain should be equal to the settings hostname.

Though there is one edge case which i'm noticing now. If there is a community with remote moderator, that remote mod wont be able to change the icon/banner (or rather, the change would not be federated to the community's instance). Not sure how to fix that.

[the remark this answers is not in the document]
Comment (dessalines) [reply crates/api_common/src/lib.rs:639]: Ohhhh okay I see now, only through the local API.

I wouldn't worry about that edge case for now.

**Review by dessalines — APPROVED** [review]

**Pull request merged**